Try to resume interrupted downloads in FileDownloader by using HTTP range requests

flashdagger opened the pull request:
Changelog: Feature: Resume interrupted file downloads if server supports it.
Docs: Omit

Close #5708
Close #3610, 

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

flashdagger committed:
if the file download is interrupted (downloaded bytes < actual size)
try to resume at the last position and append to wriiten file

retry as usual if not making any progress

flashdagger committed:
add unittest for FileDownloader

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=6791) <br/>All committers have signed the CLA.

flashdagger committed:
FileDownloader - refactor unittests and add test for checking Exception if the filedownload doesn't make any progress.

flashdagger committed:
FileDownloader - add stricter check when downloading ranges
the server sometimes doesn't respond with the "Content-Range" field
we rather throw an error then

memsharded commented on conans/client/rest/file_downloader.py:
```diff
@@ -89,21 +99,30 @@ def write_chunks(chunks, path):
+                total_length > total_downloaded_size > range_start
+                and response.headers.get("accept-ranges") == "bytes"
+            ):
+                written_chunks = self._download_file(url, auth, headers, file_path, try_resume=True)
```
This assumes an infinite number of resumed downloads, as long as it keep working, it will keep recursing itself, until it errors or finished, did I understand it correctly?

flashdagger replied:
Exactly. Since I wanted to be minimally invasive, I came up with this implementation. For uninterrupted downloads the behavior should not change. Only if the server stops sending data, we have still a chance to make progress with another request as long as we make some progress. Of course that does not cover the case, when a conan command (e.g. install) is invoked a second time.

memsharded replied:
Yes, I am totally fine to not cover the case of consecutive commands, actually when something fails, Conan usually tries to remove stale files, to avoid corrupted things in the cache.

memsharded commented on conans/client/rest/file_downloader.py:
```diff
@@ -89,21 +99,30 @@ def write_chunks(chunks, path):
             gzip = (encoding == "gzip")
 
+            # TODO: refactor Progress to allow setting an initial progress
+            progress._processed_size = range_start
```
We probably need to add some public methods to ``progress`` to provide the necessary interface

flashdagger replied:
Are you planning to do this as part of this PR?

memsharded replied:
Small contribution in my branch: https://github.com/memsharded/conan/tree/feature/download-resume-review

memsharded reviewed (changes requested):
This is looking good, it is implemented elegantly and easy to understand with that recursion.

I would say that it is missing to cover the case where downloads are done to memory directly, not to file (arg ``path`` is ``None``). Probably a unittest checking that will show a bug there (the return of the method is the actual contents of the file transferred in this case). Typically done for small Conan files only, but better be complete, just in case. Could you please have a look?

flashdagger commented:
Thanks for your feedback. For the memory case I would need to carry the content over to the next iteration/recursion. Do you want me to introduce an additional parameter?

memsharded commented:
> Thanks for your feedback. For the memory case I would need to carry the content over to the next iteration/recursion. Do you want me to introduce an additional parameter?

Actually, as a second thought, I think it is good enough to avoid the recursion with resumable download when ``path=None``. A unit test would be nice though, to make sure that it is covered, and it cannot resume the download and have corrupted data.

flashdagger commented:
I see. I'll add one or two more test cases to cover the false path of the following condition:
https://github.com/conan-io/conan/blob/6339ae68b28360be644aee8ac366910168c18e50/conans/client/rest/file_downloader.py#L48

flashdagger committed:
FileDownloader - add test cases for downloading into memory

flashdagger committed:
FileDownloader - don't try resume for downloads into memory

memsharded commented on conans/client/rest/file_downloader.py:
```diff
@@ -86,24 +95,43 @@ def write_chunks(chunks, path):
             return ret, downloaded_size
 
+        def get_total_length():
+            content_range = response.headers.get("Content-Range", "")
```
I am a bit confused about this implementation. If we sent the ``range`` header, then it is mandatory that we get the "Content-Range" response header? So we probably should be looking for the ``Content-Range`` header if and only if we are in ``try_resume=True`` or ``range_start!=0``?

memsharded commented on conans/client/rest/file_downloader.py:
```diff
@@ -86,24 +95,43 @@ def write_chunks(chunks, path):
+            content_range = response.headers.get("Content-Range", "")
+            match = re.match(r"^bytes (\d+)-(\d+)/(\d+)", content_range)
+            if match:
+                assert range_start == int(match.group(1))
```
This could be a failure in the protocol or server side, maybe launch a ``ConanException`` with an informative message of what happened? That might help us to support if some user report errors.

memsharded replied:
There is a use case where this method return ``None``, which  will produce confusing NoneType exception later, if there is no match.

memsharded commented on conans/client/rest/file_downloader.py:
```diff
@@ -86,24 +95,43 @@ def write_chunks(chunks, path):
+                assert range_start == int(match.group(1))
+                return int(match.group(3))
+            else:
+                assert range_start == 0
```
This assert would be unnecessary if we checked for the Content-Range header only if range_start != 0.

memsharded committed:
minor changes

memsharded reviewed (commented):
Hi @flashdagger 

Very nice work here, I think we are almost there. I had a couple of concerns/questions about the content length, so I have branched from your work here: https://github.com/memsharded/conan/tree/feature/download-resume-review

Please have a look to the comments and such branch (I tried to do a PR for review to your fork, but apparently I can't).

That branch also includes the minor changes to the ``Progress`` class to avoid private access.

Please let me know if that makes sense or not. Thanks!

flashdagger commented:
Hello @memsharded 
thanks for the suggestion. I'm not an HTTP protocol expert but you have to decide what you can rely on :-) Downloading files is a crucial part not only in conan and you would get soon feedback from the users if something is broken. 
Since your changes look fine to me I added them to this PR.

flashdagger committed:
FileDownloadet - add test case expecting to fail if server does not properly support content ranges

memsharded reviewed (approved):
Thanks @flashdagger !

It is looking good to me. I will ask for another review, to make sure, lets try to include this in next 1.25.

Thanks for all the efforts.

SSE4 commented on conans/client/rest/file_downloader.py:
```diff
@@ -86,24 +95,42 @@ def write_chunks(chunks, path):
+                written_chunks = self._download_file(url, auth, headers, file_path, try_resume=True)
+            elif (
+                total_downloaded_size != total_length
+                and response.headers.get("Content-Encoding") != "gzip"
```
probably, better to check if ANY content encoding is enabled, not just gzip one

SSE4 replied:
also, consider checking `Transfer-Encoding` as well

SSE4 replied:
and also probably need to disable for `Multipart` content-type as well

memsharded replied:
But those things are not related to this PR at all, they were already there, right?

SSE4 replied:
what do you mean? these are newly added lines

flashdagger replied:
Yeah, I condensed the logic from original lines 97, 98 and 106.

memsharded replied:
Yes, that is what I meant. Those lines were there before this PR, just moved. If those things need to be changed/fixed, better be done in a separate PR. Please @SSE4 open an issue to log them for later work.

flashdagger reviewed (commented):
Some feedback from my company: We tried a huge package download from our Artifactory Pro server (v6.18) via the VPN. As the connection reliably drops at 1 GB, conan was able to continue from there. 😄 
I figured out if you set the try_resume parameter to True by default it even works within the retry loop (e.g. when there is a network error). But then you have to make sure the file does not exist initially...

memsharded commented:
> Some feedback from my company: We tried a huge package download from our Artifactory Pro server (v6.18) via the VPN. As the connection reliably drops at 1 GB, conan was able to continue from there.

Excellent, great news!

> I figured out if you set the try_resume parameter to True by default it even works within the retry loop (e.g. when there is a network error). But then you have to make sure the file does not exist initially...

Oh, yes, that would make sense, it would be a problem if the file was already there (and maybe corrupted) from previous Conan runs (not the retry loop), but in theory the mechanisms (the dirty flag, and removing files) should take care of this. Do you think then it is better to activate ``try_resume=True`` and check the file existence? Would you like to contribute it now or would it be ok as-is as a first approach?

flashdagger commented:
> Do you think then it is better to activate `try_resume=True` and check the file existence? Would you like to contribute it now or would it be ok as-is as a first approach?

I'm fine with the implementation so far. If users need more robust solutions they can provide more feedback and we can work on it. What do you think?

memsharded commented:
> I'm fine with the implementation so far. If users need more robust solutions they can provide more feedback and we can work on it. What do you think?

yeah, agree. Let's wait for feedback then.

jgsogo reviewed (approved):
I've been trying to find a flaw in the implementation but it looks like any scenario I can think of will result in the same behavior as before, so 💯. Really good job here, @flashdagger
